### PR TITLE
improve router docker debug image

### PR DESCRIPTION
**Pull request by garypen**
fixes: #2135
fixes: #2145 

This PR repurposes the -debug image to make it the basis for a memory tracking image which we can use for investigating router memory issues via heaptrack. (https://github.com/KDE/heaptrack)

The PR is a *breaking* change because it now automatically starts the router under the control of heaptrack. Technically, it's not really a breaking change and it's certainly not an API change, but I really want to draw people's attention to the fact that the debug image will now execute a lot slower than the non-debug image and use a lot more memory (to track memory with...).

I've updated the docker documentation to show how to mount a local directory to store the heaptrack data. I haven't updated the kubernetes docs, because we don't go into that level of detail and we assume that a kubernetes devops person would know how to allocate and mount a PVC.

### CODE REVIEW

File: dockerfiles/Dockerfile.router [hunk 1]
```diff
@@ -1,15 +1,25 @@
...
+  if [ "${DEBUG_IMAGE}" = "true" ]; then \
+    apt-get install -y heaptrack && \
+    mkdir data; \
+  fi
+
+# Clean up apt lists
+RUN rm -rf /var/lib/apt/lists/*
 
 # Run the Router downloader which puts Router into current working directory
 RUN curl -sSL https://router.apollo.dev/download/nix/${ROUTER_RELEASE}/ | sh
```
Comment (Geal) [inline dockerfiles/Dockerfile.router:25]: the router binary that is used here will be without debug symbols, right? Will heaptrack be able to display enough information?

**Review by Geal — APPROVED** [review]

File: dockerfiles/Dockerfile.router (hunk 1 above)
Comment (garypen) [reply dockerfiles/Dockerfile.router:25]: You get the output we were looking at this morning. Things are mangled but you can roughly work out what is happening. I do think we can improve this further down the line, this is a step along the way.

File: docs/source/containerization/docker.mdx [hunk 2]
```diff
@@ -58,18 +58,32 @@ docker run -p 4000:4000 \
   --rm \
   --interactive \
   --tty \
-  --entrypoint=sh
-  ghcr.io/apollographql/router:<image version>-debug
-/dist # pwd
+  --entrypoint=bash \
+  ghcr.io/apollographql/router:<image version>
+root@0c6ae96a2bbd:/dist# pwd
```
Comment (bnjjj) [inline docs/source/containerization/docker.mdx:63]: Do we need this prefix ?

File: docs/source/containerization/docker.mdx [hunk 3]
```diff
@@ -58,18 +58,32 @@ docker run -p 4000:4000 \
...
 /dist
-/dist # ls
-LICENSE        config         router
-README.md      licenses.html  schema
-/dist # exit
+root@0c6ae96a2bbd:/dist# ls
```
Comment (bnjjj) [inline docs/source/containerization/docker.mdx:65]: Same

File: docs/source/containerization/docker.mdx [hunk 4]
```diff
@@ -58,18 +58,32 @@ docker run -p 4000:4000 \
...
+config	router	schema
+root@0c6ae96a2bbd:/dist# exit
```
Comment (bnjjj) [inline docs/source/containerization/docker.mdx:67]: same

File: docs/source/containerization/docker.mdx (hunk 2 above)
Comment (garypen) [reply docs/source/containerization/docker.mdx:63]: No. It's just to show what it would look like. I can take them out.

**Review by BrynCooke — APPROVED** [review]
Had a go looks great:
```
heaptrack stats:
	allocations:          	2036
	leaked allocations:   	202
	temporary allocations:	203
Heaptrack finished! Now run the following to investigate the data:

  heaptrack --analyze "/dist/data/router_heaptrack.gz"
```

File: docs/source/containerization/docker.mdx (hunk 2 above)
Comment (garypen) [reply docs/source/containerization/docker.mdx]: done in: https://github.com/apollographql/router/pull/2142/commits/2bd3a4e0e4758a470e6351749f2fb3d2598bf376

File: docs/source/containerization/docker.mdx (hunk 3 above)
Comment (garypen) [reply docs/source/containerization/docker.mdx]: done in: https://github.com/apollographql/router/pull/2142/commits/2bd3a4e0e4758a470e6351749f2fb3d2598bf376

File: docs/source/containerization/docker.mdx (hunk 4 above)
Comment (garypen) [reply docs/source/containerization/docker.mdx]: done in: https://github.com/apollographql/router/pull/2142/commits/2bd3a4e0e4758a470e6351749f2fb3d2598bf376

**Pull request merged**